nhais-adaptor/edifact/incoming/parser/deserialiser.py

```diff
@@ -14,48 +16,37 @@
 INTERCHANGE_TRAILER_KEY = "UNZ"
 
 
-def extract_relevant_lines(original_dict: EdifactDict, starting_pos: int, trigger_key: str) -> EdifactDict:
-    """
-    From the original dict generate a smaller dict just containing the relevant lines based upon the trigger key
-    will keep looping till the terminating key is found in the terminating config.
-    :param original_dict: The original larger dictionary.
-    :param starting_pos: The starting position to start the loop from.
-    This is to prevent starting the loop from the start each time and be slightly more efficient.
-    :param trigger_key: The trigger key for this section that will be used to find what the
-    terminating key for the section is.
-    :return: A smaller dictionary with just the relevant lines for the section.
-    """
-    terminating_config = {
-        INTERCHANGE_HEADER_KEY: [MESSAGE_HEADER_KEY],
-        MESSAGE_BEGINNING_KEY: [MESSAGE_REGISTRATION_KEY],
-        MESSAGE_REGISTRATION_KEY: [MESSAGE_PATIENT_KEY, MESSAGE_TRAILER_KEY],
-        MESSAGE_PATIENT_KEY: [MESSAGE_TRAILER_KEY]
-    }
+def deserialise_interchange_header(original_dict: EdifactDict, index: int) -> InterchangeHeader:
+    interchange_header_line = EdifactDict(helpers.extract_relevant_lines(original_dict, index, [MESSAGE_HEADER_KEY]))
+    interchange_header = creators.create_interchange_header(interchange_header_line)
+    return interchange_header
 
-    new_dict = EdifactDict([])
-    for (key, value) in original_dict[starting_pos:]:
-        if key not in terminating_config[trigger_key]:
-            new_dict.append((key, value))
-        else:
-            break
 
-    return new_dict
+def deserialise_message_beginning(original_dict: EdifactDict, index: int) -> MessageSegmentBeginningDetails:
+    msg_bgn_lines = EdifactDict(helpers.extract_relevant_lines(original_dict, index, [MESSAGE_REGISTRATION_KEY]))
+    msg_bgn_details = creators.create_message_segment_beginning(msg_bgn_lines)
+    return msg_bgn_details
 
 
-def convert_to_dict(lines: List[str]) -> EdifactDict:
-    """
-    Takes the list of original edifact lines and converts to a dict.
-    :param lines: a list of string of the original edifact lines.
-    :return: EdifactDict - A list of Tuples. Since the keys in the edifact interchange can
-    contain duplicates a tuple is required here rather than a set.
-    """
-    generated_dict = EdifactDict([])
+def deserialise_transaction(original_dict: EdifactDict, index: int) -> Transaction:
+    transaction_pat = None
+    # When extracting the relevant lines for a transactions we need to skip the very first REGISTRATION_KEY
+    # Therefore we start the index at +1.
+    transaction_lines = EdifactDict(

```

Comment:
I found this a little difficult to follow. Perhaps worth extracting the creation of the registration transaction and the patient transaction into separate methods, to make it a little clearer what's going on?

Reply:
This was a little tricky one. I've had to extract it out to quite a few separate functions. Since it needed some of the details determined before later in in this function. Let me know if this is any better?

Reply:
👍 I think it's definitely clearer.